subxt/src/backend/mod.rs

```diff
@@ -336,177 +336,293 @@ pub struct StorageResponse {
                 Box::pin(async {
                     let mut data = self.data.lock().await;
-                    let values: RpcResult<Vec<RpcResult<Box<RawValue>>>> =
-                        data.subscription.read().await.map(|v| {
-                            v.into_iter()
-                                .map(|v| {
-                                    v.map(|v| serde_json::value::RawValue::from_string(v).unwrap())
-                                })
-                                .collect::<Vec<RpcResult<Box<RawValue>>>>()
-                        });
-                    values.map(|v| RawRpcSubscription {
-                        stream: futures::stream::iter(v).boxed(),
-                        id: Some("ID".to_string()),
-                    })
+                    data.subscribe(sub, params).await.await
                 })
             }
         }
+    }
 
-        // Define dummy config
-        enum Conf {}
-        impl Config for Conf {
-            type Hash = crate::utils::H256;
-            type AccountId = crate::utils::AccountId32;
-            type Address = crate::utils::MultiAddress<Self::AccountId, ()>;
-            type Signature = crate::utils::MultiSignature;
-            type Hasher = crate::config::substrate::BlakeTwo256;
-            type Header = crate::config::substrate::SubstrateHeader<u32, Self::Hasher>;
-            type ExtrinsicParams = DefaultExtrinsicParams<Self>;
+    // Define dummy config
+    enum Conf {}
+    impl Config for Conf {
+        type Hash = crate::utils::H256;
+        type AccountId = crate::utils::AccountId32;
+        type Address = crate::utils::MultiAddress<Self::AccountId, ()>;
+        type Signature = crate::utils::MultiSignature;
+        type Hasher = crate::config::substrate::BlakeTwo256;
+        type Header = crate::config::substrate::SubstrateHeader<u32, Self::Hasher>;
+        type ExtrinsicParams = DefaultExtrinsicParams<Self>;
+        type AssetId = u32;
+    }
 
-            type AssetId = u32;
+    mod legacy {
+        use super::*;
+        use crate::backend::{
+            legacy::rpc_methods::Bytes, legacy::rpc_methods::RuntimeVersion, legacy::LegacyBackend,
+        };
+        use rpc_client::*;
+
+        pub fn setup_mock_rpc() -> MockRpcBuilder {
+            MockRpcBuilder::default()
+                .add_method(
+                    "state_getStorage",
+                    Box::new(|data, _sub, params| {
+                        Box::pin(async move {
+                            let params = params.map(|p| p.get().to_string());
+                            let rpc_params = jsonrpsee::types::Params::new(params.as_deref());
+                            let key: sp_core::Bytes = rpc_params.sequence().next().unwrap();
+                            let value = data.pop(key.0).unwrap_single();
+                            value.map(|v| serde_json::value::RawValue::from_string(v).unwrap())
+                        })
+                    }),
```

Comment:
I wonder how easily we could remove the Box's here to make the API cleaner.

Eg if `add_method` and `add_subscription` took in some generic like:

```rust
fn add_subscription(sub: &str, f: F) 
where
    Fn: Send + Fn(
        &'a mut MockDataTable,
        &'a mut Option<Subscription>,
        Option<Box<serde_json::value::RawValue>>,
    ) -> RawRpcFuture<'a, Box<serde_json::value::RawValue>>,
```

Then you could do the boxing internally and keep the API cleaner (maybe you also return some future that is then boxed or whatever instead of returning `RawRpcFuture<'a, Box<serde_json::value::RawValue>>` so that both boxes are removed?)

The downside is you probably end up having to copy these bounds in both the `Box` and in the `where` though could maybe avoid duplication with a tiny macro or trait like `trait SubscriptionFn: Fn(...)`.

Not a big deal really because this is just test code but might be a nice little ergonomic win :)

Reply:
removed boxing form the callsite